packages/common/container-definitions/src/containerExtension.ts

```diff
@@ -0,0 +1,126 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import type {
+	JsonDeserialized,
+	JsonSerializable,
+} from "@fluidframework/core-interfaces/internal";
+
+/**
+ * While connected the id of a client within a session.
+ *
+ * @internal
+ */
+export type ClientConnectionId = string;
+
+/**
+ * Common interface between incoming and outgoing signals.
+ *
+ * @internal
+ */
+export interface IExtensionMessage<TType extends string = string, TContent = unknown> {
+	/**
+	 * Message type
+	 */
+	type: TType;
+
+	/**
+	 * Message content
+	 */
+	content: JsonDeserialized<TContent>;
+
+	/**
+	 * The client ID that submitted the message.
+	 * For server generated messages the clientId will be null.
+	 */
+	// eslint-disable-next-line @rushstack/no-new-null
+	clientId: ClientConnectionId | null;
+
+	/**
+	 * Client ID of the singular client the message is being (or has been) sent to.
+	 * May only be specified when IConnect.supportedFeatures['submit_signals_v2'] is true, will throw otherwise.
+	 */
+	targetClientId?: ClientConnectionId;
+}
+
+/**
+ * @internal
+ */
+export interface IContainerExtension<TContext extends unknown[]> {
```

Comment:
I did not spend that much time building mental model, but it's relatively hard to follow the responsibilities of each of these objects, and how they are used in a workflow. Having at least some comments above each interface / class providing some info like that would be super useful

Reply:
I didn't spend time documenting the container extension part too much because I think it is everything surrounding it that matters more. Details matter so I have added more comments and hopefully the relationships are more clear now. The implementation of store in packages/runtime/container-runtime/src/containerRuntime.ts (mostly `acquireExtension` at line 4650) and the use of the feature in packages/runtime/container-runtime/src/test/containerExtensionStore.extension.example.ts's `hookItUp` are perhaps more helpful.